resolver/resolver.go

```diff
@@ -161,6 +187,30 @@ func (r *dnsResolveProber) ResolveOnce(
 	if err != nil {
 		return nil, 0, err
 	}
+	switch r.affinity {
+	case AllFamilies:
+		break
+	case PreferIPv4:
+		ip4Addresses := addresses[:0]
+		for _, address := range addresses {
+			if address.Is4() || address.Is4In6() {
```

Comment:
As far as whether an IPv4 single-stack system would handle it, I think as long as the OS and network stack have IPv6 support in them to be able to handle the address the answer is broadly yes. As for whether or not that means this conditional belongs here... Probably not. I think I added it out of an abundance of caution, but there is no evidence that I'm aware of that people are sneaking IPv4 addresses into AAAA records or anything bizarre like that.

Reply:
Because we do unwrap it using `Unmap` I'm kinda unsure if we want to remove this check. If we remove this check it feels like we should remove the `Unmap` too.

Reply:
Okay. I think it might be the most intuitive behavior (and easiest to document) to just skip this condition and skip the unmap below. That way we can simply say that "PreferIPv4" will prefer results from DNS A records and "PreferIPv6" will prefer results from DNS AAAA records, and we don't need to try to document a bunch nuance about 4-in-6 addresses.

Reply:
Surprise! We actually can't. It breaks tests, because in fact there are cases where you WILL get IPv4-embedded IPv6 addresses, and we actually hit one in our tests (thank goodness.) So not only do I need to keep this logic, I also need to test it most likely.

Why? Well, because the trick we were _previously_ relying on was the short circuit that causes resolution of an IP address to be more or less a no-op, not actually hitting either the Go or system resolver. But why? [Well](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/net/lookup.go;l=309-311)....

```go
	if ip, err := netip.ParseAddr(host); err == nil {
		return []IPAddr{{IP: IP(ip.AsSlice()).To16(), Zone: ip.Zone()}}, nil
	}
```

But why? I'm not sure. I dug unnecessarily deep and I still don't really have the answer.

Unfortunately, since this is a super common case, I think we absolutely must handle it. So... nevermind.

Reply:
Okay. So I think the only thing squirrelly that might happen is if a DNS result includes an A record _and_ an AAAA record that is a 4-in-6-style address for the same thing as the A record. We'd end up establishing double connections to it, which isn't necessarily bad and is likely pretty defensible behavior.